aiohttp/lifecycle.py

```diff
@@ -0,0 +1,109 @@
+    socket_connect_timeout = attr.ib(type=float, default=None)  # TODO
+    connection_acquiring_timeout = attr.ib(type=float, default=None)  # TODO
+    new_connection_timeout = attr.ib(type=float, default=None)  # TODO
+    http_header_timeout = attr.ib(type=float, default=None)  # TODO metadata=_tm("request_sent", "response_headers_received"))
+    response_body_timeout = attr.ib(type=float, default=None)  # TODO metadata=_tm("response_headers_received", "request_end"))
+
+    # to create a timeout specific for a single request, either
+    # - create a completely new one to overwrite the default
+    # - or use http://www.attrs.org/en/stable/api.html#attr.evolve to overwrite the defaults
+    # (maybe this should be done through either session.extend_timeout or session.replace_timeout without directly calling RequestTimeouts)
+
+
+class RequestLifecycle:
+    """ Internal class used to keep together the main dependencies used
+    at the moment of send a signal."""
+
+    def __init__(self, session, loop, trace_configs, trace_request_ctx, timeout_config):
+        self._session = session
+        self._loop = loop
+        self._trace_configs = [
+            (config, config.trace_config_ctx(trace_request_ctx=trace_request_ctx)) for config in trace_configs
+        ]
+        self._timeout_config = timeout_config
+
+        self._signal_timestamps = {}
+        self._set_timeouts = defaultdict(list)
+        self._active_timeouts = defaultdict(list)
+
+        # filter timeouts that were actually set by the user
+        if timeout_config:
+            for timeout_field in attr.fields(RequestTimeouts):
+                if timeout_field.metadata and getattr(timeout_config, timeout_field.name):
+                    self._active_timeouts[timeout_field.metadata["start"]].append(
+                        (timeout_field.metadata["end"], getattr(timeout_config, timeout_field.name))
+                    )
+
+        # create a timeout context depending on wether any timeouts were actually set
+        if self._set_timeouts:
+            self.request_timer_context = TimerContext(self._loop)
+        else:
+            self.request_timer_context = TimerNoop()
+
+        # generate send_signal methods, sending on_signal to all trace listeners and keeping track of timeouts, for all Signals
+        for signal, params_class in SIGNALS:
+            setattr(self, "send_" + signal, self._send(signal, params_class))
+
+    def _send(self, signal, params_class):
+        async def sender(*args, **kwargs):
+            # record timestamp
+            self._signal_timestamps[signal] = time.time()
+
+            # cancel all running timeouts that end with this signal
+            while self._set_timeouts[signal]:
+                timeout_handle = self._set_timeouts[signal].pop()
+                timeout_handle.cancel()
+
+            # send on_signal to all trace listeners
+            params = params_class(*args, **kwargs)
+            await asyncio.gather(
```

Comment:
default gather is kinda dangerous, after first exception the un-finished coroutines will continue executing un-parented, is this what you want?  If so I wouldn't cmt why it's ok.  I think it would be weird to return to caller and then have the signals hit ;)

Reply:
two other issues:
1. Do we need to wait for receipt of each signal, should these just all be `ensure_future`s ? If we want to receipt, does send need a timeout?  ;)
2. I don't think we want to raise out if `send` raises?

Reply:
See [my comment](https://github.com/aio-libs/aiohttp/pull/2842#issuecomment-373922955) below, I don't have a strong opinion about how exactly the signal should be sent. I just saw it as a faster version of the old `for trace: await trace.send`. There's also the `return_exceptions=False` parameter or as an alternative `asyncio.wait` with its `return_when=ALL_COMPLETED, FIRST_COMPLETED or FIRST_EXCEPTION` parameter, depending on which exact behaviour we want. If you don't want to discuss this now, we could just as well stay with the old sequential approach for this PR.